include/ginkgo/core/base/lin_op.hpp

```diff
@@ -43,6 +43,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/device_matrix_data.hpp>
 #include <ginkgo/core/base/dim.hpp>
 #include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op_helpers.hpp>
```

Comment:
I think all these changes will clash with other PRs, so perhaps it's best to leave them out.

Reply:
It reduces the duplication quite a bit, particularly for the macros for Factory parameters, so I would say it is necessary. 

But, I think the Factory parameters PR might be merged before this PR and I will rebase it after that anyway and see if these changes still make sense then.

Reply:
IIRC there I move most of the factory macros to `abstract_factory.hpp`, so that should help

Reply:
No changes in lin_op now.